.ci/test_r_package_windows.ps1

```diff
@@ -84,9 +84,18 @@ if ($env:R_MAJOR_VERSION -eq "3") {
   Check-Output $false
 }
 
+# only build vignettes for CRAN-style builds
+if ($env:R_BUILD_TYPE -eq "cran") {
+  $env:LGB_BUILD_VIGNETTES = "true"
+}
+
 $env:R_LIB_PATH = "$env:BUILD_SOURCESDIRECTORY/RLibrary" -replace '[\\]', '/'
 $env:R_LIBS = "$env:R_LIB_PATH"
-$env:PATH = "$env:RTOOLS_BIN;" + "$env:RTOOLS_MINGW_BIN;" + "$env:R_LIB_PATH/R/bin/x64;" + "$env:R_LIB_PATH/miktex/texmfs/install/miktex/bin/x64;" + $env:PATH
+# NOTE: gzip and tar are needed to create a CRAN package on Windows, but
+# some flavors of tar.exe can fail in some settings on Windows. Putting
+# the msys64 utilities at the beginning of PATH to be sure they're used
+# for that purpose.
+$env:PATH = "C:\msys64\usr\bin;" + "$env:RTOOLS_BIN;" + "$env:RTOOLS_MINGW_BIN;" + "$env:R_LIB_PATH/R/bin/x64;" + "$env:R_LIB_PATH/miktex/texmfs/install/miktex/bin/x64;" + $env:PATH
```

Comment:
@jameslamb We are removing `C:\msys64\usr\bin` from PATH with the aim to avoid weird errors like #4461.

https://github.com/microsoft/LightGBM/blob/1dbf4382fd7d8cd6ce92b53c3b84aa479aabfc19/.ci/test_r_package_windows.ps1#L52

This `MSYS` can conflict with one bundled with Rtools.

Reply:
I didn't forget about that from #4463. But my impression was that that PR was aiming to remove things that MIGHT cause problems, without knowing for sure that they would.

After extensive testing here, I really think we need the shell utilities contained there to get the version of `build-cran-package.sh` in this PR working.

Anyway, you should ignore this PR until it's ready for review. I'll convert it to draft and remove reviewers, I should have just done that from the beginning.

Reply:
It was early comment to prevent you from going a wrong way. Really glad you didn't forget about #4053.

I strongly believe that two versions of the same tool in the PATH cannot bring any benefits but only bring odd conflicts. Also, it may lead to unreliable results of our CI tests: you think that you are testing against new Rtools `x.xx` version you've just installed but actually LightGBM is being built with gcc from `C:\msys64`. I think this is very dangerous thing.

> I really think we need the shell utilities contained there ...

Are ones from Rtools broken? From the commit history I can see that the problem is with tar utility. Maybe we can install it explicitly and separately from MSYS2?

I remember I had a similar problem with `qpdf` in RGF repo and end up installing it via `pacman` but then switched to separated installation from sourceforge because R was not able to find it: https://github.com/RGF-team/rgf/pull/345/files.

Also, it is said that `tar` from Rtools has some patches:
> tar has some custom patch from BDR for backward compatibility with old rtools
    https://github.com/r-windows/rtools-base/blob/0fc6010ea37ed1ac684e36fd88736e7ba567b84c/readme.md#how-it-works


I have one idea. Could you please try not to add the whole `C:\msys64\usr\bin` to `PATH` but set only `TAR` env variable?
https://github.com/wch/r-source/blob/021672b2580c0645690b0a1184c2a25e36d0026d/src/library/utils/R/tar.R#L23
https://github.com/wch/r-source/blob/021672b2580c0645690b0a1184c2a25e36d0026d/src/library/utils/R/tar.R#L36-L40

Reply:
Good suggestion! I didn't think about looking in the R source.

I tried setting `TAR` and `R_GZIPCMD` tonight, and saw a similar result to what I saw during testing on this PR.

The R 3.6 CRAN job fails with the following error.

```text
lightgbm/
tar (child)lightgbm/build/
lightgbm/build/vignette.rds
lightgbm/cleanup
lightgbm/configure
: gzip: Cannot exec: No such file or directory
tar (child): Error is not recoverable: exiting now
tar: lightgbm_3.2.1.99.tar.gz: Cannot write: Broken pipe
tar: Child returned status 2
tar: Error is not recoverable: exiting now
Running R CMD check
```

logs: https://github.com/microsoft/LightGBM/pull/3946/checks?check_run_id=3275896515

The R 3.6 CRAN job is the only one failing like this, so I maybe the problem is in Rtools35 specifically.

@StrikerRUS what do you think about doing something like this?

```powershell
} elseif ($env:R_BUILD_TYPE -eq "cran") {

    if ($env:R_MAJOR_VERSION -eq "3") {
        $env:PATH = "C:\msys64\usr\bin;" + $env:PATH
    }

    Run-R-Code-Redirect-Stderr "result <- processx::run(command = 'sh', args = 'build-cran-package.sh', echo = TRUE, windows_verbatim_args = FALSE, error_on_status = TRUE)" ; Check-Output $?
    Remove-From-Path ".*msys64.*"
```

That would fix the R3.6 builds, and would remove the `msys64` stuff from the path before the package is built.

Reply:
@jameslamb 
> what do you think about doing something like this?

Good idea! I agree with this tradeoff.

> The R 3.6 CRAN job is the only one failing like this, so maybe the problem is in Rtools35 specifically.

Seems you are right.
https://github.com/r-lib/actions/issues/13#issuecomment-558875616
> It seems to be an issue with the tar executable bundled with Rtools, it doesn't find the gzip executable when trying to compress the files even if it is on the PATH.
   https://github.com/r-lib/actions/issues/13#issuecomment-559183309

Probably setting `TAR` env variable is not enough to overwrite bundled `tar.exe`...

---

*offtopic*: I wanted to explore Windows image we work with more precisely with RDP but found out it is not allowed. The repo with RDP Workflow example was blocked.
https://github.com/nelsonjchen/reverse-rdp-windows-github-actions
https://github.community/t/is-it-legal-windows-rdp-on-github-actions/184252